### ISSUE TITLE
Image in portable text with gatsby-image, low res images get upscaled instead of shown in original size.

### ISSUE THREAD

**Issue by Ylvisthecool** [sanity-io/gatsby-source-sanity#33]
I'm trying to display images added in portable text, but so far I've found problems with both `getFluidGatsbyImage` and `getFixedGatsbyImage` that prevent me from displaying them in a desirable way.

What I really want to achieve is to set a max width, for example 800px, for large images, but display images with a width lower than 800px at their original size.

However, both `getFluidGatsbyImage` and `getFixedGatsbyImage` seem to ignore the original width and render the image at maxWidth for fluid or 400px which is the default width for fixed.

Looking through the json that is returned from the `_rawBody` response I cannot even find any metadata that tells me what the original size of the image. If I had this I could at least limit the image container to that size to circumvent the issue.

Example of my image renderer:
```javascript
import Image from "gatsby-image"
import { getFluidGatsbyImage } from "gatsby-source-sanity"
import React from "react"
import clientConfig from "../../client-config"

export function Figure({ node }) {
  const fluidProps = getFluidGatsbyImage(
    node.asset._id,
    { maxWidth: 800 },
    clientConfig.sanity
  )

  return (
    <figure>
      <Image fluid={fluidProps} alt={node.alt} />
      {node.caption && <figcaption>{node.caption}</figcaption>}
    </figure>
  )
}
```

**Comment by rexxars** [comment]
Can you try v5.0.2 and see if that solves your issue?

**Comment by hcavalieri** [comment]
**📌 long and thorough explanation incoming, sorry for the huge answer!**

I went through the same problem today when trying to get a `srcset` for a full-width image for a website hero: I'd run `fluid(maxWidth: 4000)` and get back a srcset with a single value of `1000w`, which I found very odd.

Looking into it, I realized the original image was around `1800px` wide, and I figured that `getFluidGatsbyImage` was doing something wrong... I had looked into its source code for inspiration on an implementation of a `GatsbyImage + Sanity` equivalent outside of Gatsby, and had a hunch of what was going on...

Turns out the widths are generated from a fixed set of multipliers:

https://github.com/sanity-io/gatsby-source-sanity/blob/d4e19e767ac690110e103bed7105ad68ea9c78a9/src/images/getGatsbyImageProps.ts#L82

Which is used to generate an array of widths for the image:

https://github.com/sanity-io/gatsby-source-sanity/blob/d4e19e767ac690110e103bed7105ad68ea9c78a9/src/images/getGatsbyImageProps.ts#L249-L252

The problem with this approach is that, by multiplying `maxWidth * scale`, we're going to get values that are often bigger than the image's original width. **Picture this** (_pun intended_ 😝): someone is building a page that has a section with up to `1600px` in width, of which an image occupies the whole background. The developer, seeing this, will get the asset from the design handoff, without taking much notice that it's actually `1500px` wide, and upload it to Sanity. Then, in the GraphQL query they'll add `fluid(maxWidth: 1600)`, which is absolutely expected, given the design constraint, but the result won't be what they expect:

```js
const widths = [0.25, 0.5, 1, 1.5, 2, 3].map(
     scale => scale * 1600 // desired maxWidth
   )
   .filter(width => width < 1500 // original image's dimension)

// result: 375w, 750w
```

The **final result will be a blurry `750px` picture in a `1600px` container**, which will be hard to debug and which demands a bit of knowledge on `srcset` and the `picture` element. We can deal with this in several ways:

1. completely ignoring the array `filter` and sending widths that aren't realistic for given images. I think this is the weakest solution;
1. tweaking `sizeMultipliersFluid` and adding extra values. Doesn't _solve_ the problem, but is a good remedy for a patch release;
1. through warnings by logging to the Gatsby `reporter` if a given fluid image comes out at a width considerably smaller than the desired `maxWidth` property. Could become annoying on the console, possibly requiring a `verbose` boolean flag in the plugin options; and still doesn't fix the problem.
1. and, finally, the one I suggest us doing: instead of working with multipliers, we could work with a `minimalWidth` and a `widthStep` combination. That'd create an array of widths going from the smallest value possible all the way to the desired `maxWidth`, and then, if the file supports it, we could add multipliers up to `3x` for displaying crisp images for high DPI displays 😄

That was my approach after much tinkering around, and I'm very happy with it! The drawback is having potentially huge `srcset`s, but if we go the extra mile we could cap this somehow 🤔 Also, you might want to take a look at [this Smashing Magazine article](https://www.smashingmagazine.com/2016/09/automating-art-direction-with-the-responsive-image-breakpoints-generator/) by a dev. advocate from Cloudinary evaluating possible approaches. He concludes that **the most important factor for creating a srcset is to consider the bandwidth gains** of providing smaller sizes. It's a tradeoff between pixels/resolution and download size, and he finishes by introducing Cloudinary's automatic breakpoint generator, a nice service that we unfortunately can't use for this project... So the `px` stepping is the second best option, IMO.

I'd gladly help with the implementation, or I could open source the code I did as well, its a bit rough and doesn't cover every use case, but it's a start ✨
